language/move-stdlib/tests/VectorTests.move

```diff
@@ -0,0 +1,398 @@
+        assert(*V::borrow(&v, 2) == 2, 3);
+        assert(*V::borrow(&v, 3) == 3, 4);
+
+        V::reverse(&mut v);
+
+        assert(*V::borrow(&v, 3) == 0, 5);
+        assert(*V::borrow(&v, 2) == 1, 6);
+        assert(*V::borrow(&v, 1) == 2, 7);
+        assert(*V::borrow(&v, 0) == 3, 8);
+    }
+
+    #[test]
+    fun reverse_vector_nonempty_odd_length_non_singleton() {
+        let v = V::empty();
+        V::push_back(&mut v, 0);
+        V::push_back(&mut v, 1);
+        V::push_back(&mut v, 2);
+
+        assert(*V::borrow(&v, 0) == 0, 1);
+        assert(*V::borrow(&v, 1) == 1, 2);
+        assert(*V::borrow(&v, 2) == 2, 3);
+
+        V::reverse(&mut v);
+
+        assert(*V::borrow(&v, 2) == 0, 4);
+        assert(*V::borrow(&v, 1) == 1, 5);
+        assert(*V::borrow(&v, 0) == 2, 6);
+    }
+
+    #[test]
+    #[expected_failure(abort_code = 1)]
+    fun swap_empty() {
+        let v = V::empty<u64>();
+        V::swap(&mut v, 0, 0);
+    }
+
+    #[test]
+    #[expected_failure(abort_code = 1)]
+    fun swap_out_of_range() {
+        let v = V::empty<u64>();
+
+        V::push_back(&mut v, 0);
+        V::push_back(&mut v, 1);
+        V::push_back(&mut v, 2);
+        V::push_back(&mut v, 3);
+
+        V::swap(&mut v, 1, 10);
+    }
+
+    #[test]
+    #[expected_failure(abort_code = 0)]
+    fun swap_remove_empty() {
+        let v = V::empty<u64>();
+        V::swap_remove(&mut v, 0);
+    }
+
+    #[test]
+    fun swap_remove_singleton() {
+        let v = V::empty<u64>();
+        V::push_back(&mut v, 0);
+        assert(V::swap_remove(&mut v, 0) == 0, 0);
+        assert(V::is_empty(&v), 1);
+    }
+
+    #[test]
+    fun swap_remove_inside_vector() {
+        let v = V::empty();
+        V::push_back(&mut v, 0);
+        V::push_back(&mut v, 1);
+        V::push_back(&mut v, 2);
+        V::push_back(&mut v, 3);
+
+        assert(*V::borrow(&v, 0) == 0, 1);
+        assert(*V::borrow(&v, 1) == 1, 2);
+        assert(*V::borrow(&v, 2) == 2, 3);
+        assert(*V::borrow(&v, 3) == 3, 4);
+
+        assert(V::swap_remove(&mut v, 1) == 1, 5);
+        assert(V::length(&v) == 3, 6);
+
+        assert(*V::borrow(&v, 0) == 0, 7);
+        assert(*V::borrow(&v, 1) == 3, 8);
+        assert(*V::borrow(&v, 2) == 2, 9);
+
+    }
+
+    #[test]
+    fun swap_remove_end_of_vector() {
+        let v = V::empty();
+        V::push_back(&mut v, 0);
+        V::push_back(&mut v, 1);
+        V::push_back(&mut v, 2);
+        V::push_back(&mut v, 3);
+
+        assert(*V::borrow(&v, 0) == 0, 1);
+        assert(*V::borrow(&v, 1) == 1, 2);
+        assert(*V::borrow(&v, 2) == 2, 3);
+        assert(*V::borrow(&v, 3) == 3, 4);
+
+        assert(V::swap_remove(&mut v, 3) == 3, 5);
+        assert(V::length(&v) == 3, 6);
+
+        assert(*V::borrow(&v, 0) == 0, 7);
+        assert(*V::borrow(&v, 1) == 1, 8);
+        assert(*V::borrow(&v, 2) == 2, 9);
+    }
+
+    #[test]
+    #[expected_failure(abort_code = 1)]
```

Comment:
whats error code 1? lol 

I guess we can't do `Vector::ERROR_CODE` or something?

Reply:
Sadly not :( This and being able to access private functions/construct data in the tested module are the main reasons for why we might want an extension module vs. a module with `#[test_only]` eventually, but there are ways to get around this with judicious use of `#[test_only]` :)